### ISSUE TITLE
unexpected output with pbs-attn

### ISSUE THREAD

**Issue by mikeallen39** [xinghaow99/pbs-attn#1]
Hi PBS-Attn Team,

First of all, thank you so much for your excellent work on pbs-attn and for making it publicly available!

I’ve been experimenting with your implementation using `pbs-attn/pbs_attn/patch/huggingface.py`, and I ran a test with a fairly long and complex prompt (see below). However, when I run without pbs-attn, the output is a reasonable continuation of the prompt(`This question invites a comprehensive exploration of the state-of-the-art in transformer-based architectures, highlighting the interplay between architectural innovations, training paradig`). However, when I apply the pbs-attn, the generated output is a long sequence of exclamation marks (`!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!`), which seems unusual and suggests something may be going wrong during generation.

Here’s the prompt I used:

> In the rapidly evolving landscape of artificial intelligence, particularly with the advent of foundation models and large-scale multimodal systems, how do contemporary transformer-based architectures reconcile the inherent tension between model capacity, computational efficiency, and generalization across diverse tasks—ranging from low-resource language translation and medical image analysis to real-time autonomous decision-making in robotics?  
    Specifically, what roles do architectural innovations such as sparse attention mechanisms (e.g., Longformer, BigBird), mixture-of-experts (MoE) layers, adaptive computation paths, and cross-modal alignment strategies (as seen in models like Flamingo or LLaVA) play in enabling scalable, energy-efficient, and robust inference without catastrophic forgetting or domain collapse?  
    Furthermore, how do training paradigms—including self-supervised pretraining on web-scale corpora, instruction tuning with human feedback (e.g., RLHF or DPO), and continual or lifelong learning frameworks—interact with these architectural choices to shape the emergent capabilities, safety properties, and alignment with human intent in deployed AI systems?  
    Finally, what theoretical or empirical limitations currently hinder the seamless integration of symbolic reasoning, causal inference, and world knowledge into end-to-end differentiable models, and what promising research directions—such as neuro-symbolic hybrids, retrieval-augmented generation, or energy-based modeling—might bridge this gap in the next generation of AI architectures?

Thanks again for your time!

**Comment by xinghaow99** [comment]
Hi @mikeallen39,

Thanks for trying out PBS-Attn! I tried to reproduce this with your prompt but got an identical generation with standard SDPA.

Can you share more information about your environment or setup so I can look further into it?

**Comment by mikeallen39** [comment]
Thank you very much for your reply!

Here is my current environment:

```
accelerate               1.10.1
aiohappyeyeballs         2.6.1
aiohttp                  3.13.0
aiosignal                1.4.0
annotated-types          0.7.0
antlr4-python3-runtime   4.9.3
anyio                    4.11.0
async-timeout            5.0.1
attrs                    25.4.0
av                       13.1.0
beautifulsoup4           4.13.4
bidict                   0.23.1
blinker                  1.9.0
block_sparse_attn        0.0.1
certifi                  2025.10.5
cffi                     2.0.0
charset-normalizer       3.4.3
click                    8.3.0
clip                     1.0
coloredlogs              15.0.1
contourpy                1.3.2
cryptography             46.0.2
cycler                   0.12.1
dashscope                1.24.6
datasets                 3.2.0
diffusers                0.31.0
dill                     0.3.8
dominate                 2.9.1
easydict                 1.13
einops                   0.8.1
exceptiongroup           1.3.0
ffmpeg-python            0.2.0
filelock                 3.13.1
flash_attn               2.8.3
flashinfer               0.1.6+cu121torch2.4
Flask                    3.1.2
Flask-SocketIO           5.5.1
flatbuffers              25.9.23
fonttools                4.60.1
frozenlist               1.8.0
fsspec                   2024.6.1
ftfy                     6.3.1
future                   1.0.0
gitdb                    4.0.12
GitPython                3.1.45
h11                      0.16.0
hf-xet                   1.2.0
huggingface-hub          0.36.0
humanfriendly            10.0
idna                     3.10
imageio                  2.34.0
imageio-ffmpeg           0.6.0
importlib_metadata       8.7.0
inquirerpy               0.3.4
itsdangerous             2.2.0
Jinja2                   3.1.4
kiwisolver               1.4.9
lazy_loader              0.4
lmdb                     1.7.3
loguru                   0.7.3
Mako                     1.3.10
MarkupSafe               2.1.5
matplotlib               3.10.7
minference               0.1.6.0
ml_dtypes                0.5.3
mpmath                   1.3.0
multidict                6.7.0
multiprocess             0.70.16
networkx                 3.3
ninja                    1.13.0
numpy                    1.24.4
nvidia-cublas-cu12       12.4.5.8
nvidia-cuda-cupti-cu12   12.4.127
nvidia-cuda-nvrtc-cu12   12.4.127
nvidia-cuda-runtime-cu12 12.4.127
nvidia-cuda-runtime-cu13 0.0.0a0
nvidia-cudnn-cu12        9.1.0.70
nvidia-cufft-cu12        11.2.1.3
nvidia-curand-cu12       10.3.5.147
nvidia-cusolver-cu12     11.6.1.9
nvidia-cusparse-cu12     12.3.1.170
nvidia-cusparselt-cu12   0.6.2
nvidia-ml-py             12.575.51
nvidia-nccl-cu12         2.21.5
nvidia-nvjitlink-cu12    12.4.127
nvidia-nvtx-cu12         12.4.127
nvidia-pyindex           1.0.9
nvidia-tensorrt          99.0.0
nvitop                   1.5.1
omegaconf                2.3.0
onnx                     1.19.1
onnx-ir                  0.1.10
onnxconverter-common     1.16.0
onnxruntime              1.23.1
onnxscript               0.5.3
open_clip_torch          3.2.0
opencv-python            4.11.0.86
packaging                25.0
pandas                   2.3.3
pbs_attn                 0.1.0          
pfzy                     0.3.4
pillow                   11.0.0
pip                      25.2
platformdirs             4.5.0
prompt_toolkit           3.0.52
propcache                0.4.1
protobuf                 6.32.1
psutil                   7.1.0
pyarrow                  21.0.0
pycocotools              2.0.10
pycparser                2.23
pycuda                   2025.1.2
pydantic                 2.10.6
pydantic_core            2.27.2
pyparsing                3.2.5
python-dateutil          2.9.0.post0
python-engineio          4.12.3
python-socketio          5.14.1
pytools                  2025.2.5
pytz                     2025.2
PyYAML                   6.0.3
regex                    2025.9.18
requests                 2.32.5
safetensors              0.6.2
scikit-image             0.25.2
scipy                    1.15.3
seaborn                  0.13.2
sentencepiece            0.2.1
sentry-sdk               2.41.0
setuptools               80.9.0
simple-websocket         1.1.0
siphash24                1.8
six                      1.17.0
smmap                    5.0.2
sniffio                  1.3.1
soupsieve                2.8
starlette                0.48.0
sympy                    1.13.1
tensorrt                 10.13.3.9
tensorrt_cu13            10.13.3.9
tensorrt_cu13_bindings   10.13.3.9
tensorrt_cu13_libs       10.13.3.9
tifffile                 2025.5.10
timm                     1.0.20
tokenizers               0.22.1
torch                    2.6.0+cu124
torchao                  0.13.0
torchaudio               2.6.0+cu124
torchvision              0.21.0+cu124
tqdm                     4.67.1
transformers             4.57.1
triton                   3.2.0
typing_extensions        4.15.0
typing-inspection        0.4.2
tzdata                   2025.2
urllib3                  2.5.0
wandb                    0.22.2
wcwidth                  0.2.14
websocket-client         1.9.0
Werkzeug                 3.1.3
wheel                    0.45.1
wsproto                  1.2.0
xattn                    0.0.1   
xxhash                   3.6.0
yarl                     1.22.0
zipp                     3.23.0
zstandard                0.25.0
```

Additionally, I’m using the following system configuration:  
- **GPU**: NVIDIA A800 80GB PCIe  
- **NVIDIA Driver Version**: 550.144.03  
- **CUDA Toolkit Version**: 12.4  
- **NVIDIA-SMI**: 550.144.03

Please let me know if you need any further information. I appreciate your support very much!

**Comment by xinghaow99** [comment]
In the meantime, can you also try setting `use_triton=False` and see if the results are correct?

**Comment by mikeallen39** [comment]
I just re-ran the test and—quite surprisingly—the output is now working perfectly as expected. I really appreciate your support!

That said, I have a follow-up question regarding the paper: I’m still a bit unclear on how segment reordering preserves the autoregressive property. Would you be kind enough to explain this in a bit more detail?

Thanks again for your patience and guidance—I truly appreciate it!

**Comment by xinghaow99** [comment]
Thank you for the interest. The permutation happens within each segment so that the inter-segment causality is preserved. For example, a query token in Seg_{i+1} can safely attend to all key tokens in Seg_i.

**Comment by mikeallen39** [comment]
However, the order of tokens within Seg_i has been changed—doesn't this have any effect?

**Comment by xinghaow99** [comment]
The intra-segment causality is ensured by a token-level mask, which corresponds to the green blocks in Fig.1 in the paper.
If you look into the implementation of Flash Attention, you'll see three types of blocks:
(a) blocks under the diagonal (yellow); (b) blocks on the diagonal (green); (c) blocks above the diagonal (grey).
(a) blocks can be computed with full attention, due to the inter-block causality constraint (similar to our inter-segment constraint here); (b) blocks get computed with a token-level mask, typically a diagonal causal mask; (c) blocks get skipped.

I hope this helps.

**Comment by mikeallen39** [comment]
Thank you so much for your patient reply! I’ll continue reading the paper and code carefully. This is truly nice work!
